Resolved Silver issue #586

dewert99 opened the pull request:
Fixes #586

dewert99 committed:
Fixed issue where magic wands with differently named let bound variables were treated as different

dewert99 committed:
Merge remote-tracking branch 'upstream/master' into wand-let

alexanderjsummers reviewed (commented):
This looks good to me - indeed, let expressions should be treated analogously to quantified variables (inside wands).

Minor but could you add to the test a case that is supposed to fail (with appropriate test annotations)? Maybe mixing two lets/quantifiers in different orders, but renaming them so that syntactically the inner assertions look identical?

Aurel300 commented on src/test/resources/all/issues/silver/0586.vpr:
```diff
@@ -0,0 +1,17 @@
+method test2(head: Ref)
+{
+	package (let x == (0) in acc(head.val)) --* true
+	exhale  (let x == (0) in acc(head.val)) --* true
```
Should the second `x` here not be e.g. `y` to test the problem you point out in #586?

dewert99 committed:
Added intentionally failing test

dewert99 committed:
Merge branch 'wand-let-old' into wand-let

dewert99 commented:
I've fixed the 'x' vs 'y' issue, and added an intentionally failing test.
I didn't give a test that mixes let and forall because
```
method test3(xs: Set[Bool]) {
  package (forall b1: Bool :: let ys == (xs) in (b1 in ys ==> true)) --* true
}
```
Gives `java.lang.RuntimeException: Internal Error: variable ys is not defined.` independently of the changes made here which should probably be it's own issue
Edit: The error is only in Carbon

alexanderjsummers reviewed (approved):
This looks good to me, thanks!

Please do file a Carbon issue for the crash you found when mixing binders.
